### PR TITLE
make improvements of ceph

### PR DESCRIPTION
**Pull request by chengtcli**
1. Run ceph-osd in parallel to speed up ursula.
2. Make ceph-update only triggered by ceph config file changed or ceph package upgrade, scale-out will not trigger ceph-update
3. Detect scale-out by running_osd_hosts_num < groups['ceph_osds']|length, don't need `ceph.scale-out` extra variable any more.
4. Accept `near full osd` warning of ceph too. We only accept health_ok before.
5. Add a new module to persist variables those are used to trigger service restart.
6. I will update the guide to notice that we should never do ceph scale-out together with ceph package upgrade or ceph config file changed. In case the operator may miss this notice, we tolerate this misoperation in ceph-update role by waiting for ceph rebalancing for top to 6 hours.

restarting ceph services should only run with ceph package upgrade or
ceph config file update. scale-out shouldn't trigger ceph service
restarting, and we have to keep in mind that we can't scale-out together
with ceph package upgrade or ceph config file update.

It may take hours to rebalance after scale-out, ceph health is not ok until rebalance is done. While we can't do ceph-update(restarting ceph services) until ceph health ok.

My design is like this:
We never do scale-out together with changing ceph.conf or upgrading ceph.
If new osd nodes are in other pod, we can enable multi-pod first. Then do scale-out.
I will update scale-out document if this design is accepted.

### CODE REVIEW

**Comment by chengtcli** [comment]
tested in tardis by using VMs and redhat.
deploy env: envs/example/ci-ceph-redhat
result:
deployment is finished without errors.
osd nodes are deployed in parallel.
Create VM and volume successfully, and is able to attach volume to VM.

File: roles/ceph-osd/tasks/main.yml [hunk 1]
```diff
@@ -2,8 +2,13 @@
...
 - include: pool_names.yml
   delegate_to: "{{ groups['ceph_monitors'][0] }}"
-  when: inventory_hostname == groups['ceph_osds'][0]
-  tags: ['monitoring', 'openstack', 'control', 'glance', 'cinder', 'cinder-data', 'nova', 'nova-data', 'data', 'openstack-setup']
+  # there is a bug with ansible
+  # https://github.com/ansible/ansible/issues/15365
+  # delegate_facts does not work with run_once
+  # so we use inventory_hostname == play_hosts[0] instead of run_once
+  when: inventory_hostname == play_hosts[0]
+  tags: ['monitoring', 'openstack', 'control', 'glance', 'cinder',
```
Comment (nirajdp76) [inline roles/ceph-osd/tasks/main.yml]: Do we still this tags here?

**Comment by nirajdp76** [comment]
Have several questions:
1: How are we protecting in code around "never do scale-out and multi-pod together"? i.e we want to scale out ceph and SL gave that node in different pod. 
We need documentation, but operator skipping doc and do end up ceph scale-out with multi-pod (which will change ceph.conf on existing ceph osd nodes as it needs to add 2nd pod cidrs), will lead ceph to go in Error state depending on how the ursula was run.

I think there is `-e update_routes=yes` needs to be passed for multi-pod expansion. So we could add some check in preflight to fail if we adding new ceph nodes when update_routes=yes
Now, we need to improve flight check for ceph health as well .. Today it will fail wuth ceph is in warning, but typically when we scale out ceph we would be in warning and scaling out to get back to health ok. At same time, when we doing regular cloud update/upgrade we do want to fail if ceph health is in warning.

2: We ur proposal, we really do not have any need for ceph.scaleout. Ceph expansion would be similar to how we expand compute nodes.. As u said doc needs to be updated

3: When updating docs, we need to make WIP PR on doc repo. This changed doc should not be merged until we GA. Also, we need to highlight in BOLD (big letters) that this is applicable to new release only. 

4: Are you going to work in differnet PR on add smart to ceph-update role for restarting osd's  / mon's?  i.e instead of relying on facts (ceph_update), we look at running process version and installed ceph version and if they are different we need to restart osd .. i,e there is no inter-decencies on main ceph rols and ceph-update. 

5: chown tasks in ceph-update .. we only need this when going from pre jewel to jewel ..,we should again remove any dependcies on set_facts and just rely on running process version and install version and wrap these expensive chown tasks around running process less than jewel and install version jewel or grate

**Comment by chengtcli** [comment]
@nirajdp76 , let me answer your questions.
1.  Thanks for remind of protecting in code from operator's misoperation.
To prevent operator misoperation, first, we have to let code know if ceph.conf updated, if ceph package is upgraded, if we are doing scale-out. But ursula code will not know if ceph.conf is updated until it run task to try to update ceph.conf. ursula code will not know if ceph package is upgraded until ceph package is really upgraded.
If the operator really missed our notes in document and do scale-out together with ceph.conf update, the only way what I can think of to prevent going error is to dynamically increase time_wait up to hours in ceph-update. This will make ursula taking more hours, but it's deserve for preventing error, right?
My initial though is to let operator skip ceph precheck by --skip-tags, I think your advice of improving precheck is better. I will do that.
2. right, I will update related documents.
3. okay, thanks for reminds
4. yes, and make ceph-config separated from ceph-common role, as we talked.
5. got it. Will do it in next PR.

**Comment by chengtcli** [comment]
PR updated, and as well the PR description. @nirajdp76 could you help review this RR again? Thanks.

This PR is tested on tardis by using VM,
deployment envs dir: envs/example/ci-ceph-redhat
tested cases:
1. run deployment. It is finished without errors.
2. check ursula.log. osd nodes are deployed in parallel.
3. create VM and volume successfully, and I can attach volume to VM.
4. chang ceph.conf by adding useless comment lines. Then rerun ursula, and manually stop ursula before ceph-update. Then rerun ursula again, ceph-update is triggered even not ceph.conf update this time.

**Comment by nirajdp76** [comment]
ok to test

File: roles/ceph-osd/tasks/main.yml [hunk 2]
```diff
@@ -2,8 +2,13 @@
...
+         'cinder-data', 'nova', 'nova-data', 'data', 'openstack-setup']
```
Comment (knandya) [inline roles/ceph-osd/tasks/main.yml]: @chengtcli : Still need to support all these tags?

File: site.yml [hunk 3]
```diff
@@ -187,7 +187,6 @@
 
 - name: ceph osds
   hosts: ceph_osds
-  serial: 1
```
Comment (knandya) [inline site.yml]: @chengtcli : Need to unit test this change with failed osd disk replacement scenario.

File: roles/preflight-checks/tasks/ceph.yml [hunk 4]
```diff
@@ -26,3 +21,17 @@
   delegate_to: "{{ item }}"
   with_items:
     - "{{ groups['controller'] }}"
+
+- name: detect scale-out
```
Comment (knandya) [inline roles/preflight-checks/tasks/ceph.yml]: @chengtcli : What would happen if preflight checks are skipped during scaleout?

File: roles/ceph-osd/tasks/main.yml (hunk 2 above)
Comment (chengtcli) [inline roles/ceph-osd/tasks/main.yml]: right, we need variables defined in pool_names.yml if we run with any of these tags

File: roles/preflight-checks/tasks/ceph.yml (hunk 4 above)
Comment (chengtcli) [inline roles/preflight-checks/tasks/ceph.yml]: We never skip precheck, we accept `HEALTHER_WARN x near full osd(s)` now

File: site.yml (hunk 3 above)
Comment (chengtcli) [inline site.yml]: right. There are many cases I will have to cover. But I would like to do full case tests after optimization.
Currently, I only cover the base cases.

Comment (knandya) [inline site.yml]: ok. Is it possible to cover the case before merge?

File: roles/preflight-checks/tasks/ceph.yml (hunk 4 above)
Comment (knandya) [inline roles/preflight-checks/tasks/ceph.yml]: Can operator skip the tests using --skip-tags? Is it going to cause problem in such case? Does it need to be documented?

File: roles/ceph-osd/tasks/main.yml (hunk 2 above)
Comment (knandya) [inline roles/ceph-osd/tasks/main.yml]: @nirajdp76 : I remember you had some concerns with the number of tags in this file. Any comments?

File: roles/preflight-checks/tasks/ceph.yml (hunk 4 above)
Comment (chengtcli) [inline roles/preflight-checks/tasks/ceph.yml]: no one should skip precheck except the developers those who know the details they are skipping.
But if really the operator skipped precheck by mistake, the ceph_scale_out is default(false)

File: site.yml (hunk 3 above)
Comment (chengtcli) [inline site.yml]: there are many cases. If I make full case tests now, I will have to do it again before optimization PR.
I would like to do full case tests before the last feature PR.
For now, I just cover the base cases.

**Comment by chengtcli** [comment]
tested on tardis by using vm
ceph-osd role takes 80-100 seconds for each osd node when running in serial.
while it takes 135 seconds to deploy 3 osd nodes in prarallel. ursula.log shows that it takes about 8 seconds for each extra osd node. So, if we have 5 osd nodes, it takes about 135 + (5-3)*8 = 151 seconds

**Comment by knandya** [comment]
retest

File: roles/ceph-osd/tasks/main.yml (hunk 2 above)
Comment (nirajdp76) [inline roles/ceph-osd/tasks/main.yml]: in 3.1.1.2 we cleared up lot of dependencies on pool_names.. Where all outside ceph-osd pool_names are needed? Also, have we ensure 3.1.1.2 cleanup PR is rolled into this where applicable.

File: site.yml (hunk 3 above)
Comment (nirajdp76) [inline site.yml]: yes, separate PR's makes more sense.

File: roles/ceph-osd/tasks/main.yml (hunk 2 above)
Comment (chengtcli) [inline roles/ceph-osd/tasks/main.yml]: yes, I have gone through all commits in 3.1 branch. And I will bring in ceph related commits which are needed in master branch, in my future PRs.

**Comment by nirajdp76** [comment]
retest

**Comment by nirajdp76** [comment]
Given there are other 3 PR lined up for ceph cleanup and optimization. This one LGTM.

**Comment by nirajdp76** [comment]
retest

**Comment by nirajdp76** [comment]
retest

**Comment by nirajdp76** [comment]
retest

**Comment by nirajdp76** [comment]
retest

**Comment by nirajdp76** [comment]
retest

**Comment by nirajdp76** [comment]
retest

**Comment by nirajdp76** [comment]
retest

**Comment by nirajdp76** [comment]
ok to test

**Comment by nirajdp76** [comment]
retest

**Comment by nirajdp76** [comment]
ok to test

**Pull request merged**